Dev desktopvirtualization microsoft.desktop virtualization 2021 05 13 preview

datr93 opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [x] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

datr93 committed:
Adds base for updating Microsoft.DesktopVirtualization from version preview/2021-04-01-preview to version 2021-05-13-preview

datr93 committed:
Updates readme

datr93 committed:
Updates API version in new specs and examples

datr93 committed:
add SAU and HPUpdate

datr93 commented:
@zhenglaizhang can you please take a look

zhenglaizhang commented:
@datr93 All checks are passed, generally it looks good for me. It is better if some guy act as code owner or are familiar with the swagger could add approving review. Currently I cannot merge it as the PR requires 1 approving review.

sampa-msft commented:
Approved

sampa-msft commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3460,6 +3460,14 @@
           "type": "string",
           "description": "Enabled allows this resource to be accessed from both public and private networks, Disabled allows this resource to only be accessed via private endpoints",
           "x-nullable": false
+        },
```
Good job

datr93 commented:
do not merge!

shailahirji commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,5668 @@
+        }
+      }
+    },
+    "SessionHostConfigurationProperties": {
```
As off yesterday need the LastUpdated timestamp in session host config

shailahirji replied:
@datr93

datr93 replied:
added the property at the same level as shc because we need to query against it.

datr93 committed:
new routes and sh props, and shc last updated time
fix errors
update SAU
corect name sessionHostConfigurationLastUpdateTime and maintenance alert array
add faultType

datr93 commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+        },
+        "primaryWindow": {
+          "$ref": "#/definitions/MaintenanceWindowProperties",
+          "description": "Primary Window of the maintenance. Maintenance windows are 2 hours long. We try to push agent update in this window first."
```
Primary Window of the maintenance. Maintenance windows are 2 hours long. We try to push component update in this window first.

datr93 replied:
done

datr93 commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+              },
+              {
+                "value": "TenantAdminControlled",
+                "description": "TenantAdmin have opted in for Scheduled Agent Update feature."
```
Scheduled Component Update feature

datr93 replied:
done

datr93 commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,6409 @@
+          "description": "The maximum virtual machines to be removed during hostpool update."
+        },
+        "maintenanceAlerts": {
+          "$ref": "#/definitions/MaintenanceAlertsProperties",
```
make a list

dallinlmag commented:
@zhenglaizhang
The new breaking changes are in a private preview product. We are trying to change the format of a string from a datetime object to match a HH:mm pattern. The properties do not need a date, just the time. I know normally, this should be a breaking change, but we don't have customers using the datetime format currently. What do you suggest?

dallinlmag commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/examples/ScalingPlan_Create.json:
```diff
@@ -0,0 +1,163 @@
+              "Thursday",
+              "Friday"
+            ],
+            "rampUpStartTime": "11:11",
+            "rampUpLoadBalancingAlgorithm": "DepthFirst",
+            "rampUpMinimumHostsPct": 20,
+            "rampUpCapacityThresholdPct": 80,
+            "peakStartTime": "11:11",
+            "peakLoadBalancingAlgorithm": "BreadthFirst",
+            "rampDownStartTime": "11:11",
+            "rampDownLoadBalancingAlgorithm": "DepthFirst",
+            "rampDownMinimumHostsPct": 20,
+            "rampDownCapacityThresholdPct": 50,
+            "rampDownForceLogoffUsers": true,
+            "rampDownWaitTimeMinutes": 30,
+            "rampDownNotificationMessage": "message",
+            "offPeakStartTime": "11:11",
+            "offPeakLoadBalancingAlgorithm": "DepthFirst"
```
@datr93 Thanks for changing the time format. I think it would make sense to have time values the customers might use, for example:
```
            "rampUpStartTime": "06:00",
            "rampUpLoadBalancingAlgorithm": "DepthFirst",
            "rampUpMinimumHostsPct": 20,
            "rampUpCapacityThresholdPct": 80,
            "peakStartTime": "08:00",
            "peakLoadBalancingAlgorithm": "BreadthFirst",
            "rampDownStartTime": "18:00",
            "rampDownLoadBalancingAlgorithm": "DepthFirst",
            "rampDownMinimumHostsPct": 20,
            "rampDownCapacityThresholdPct": 50,
            "rampDownForceLogoffUsers": true,
            "rampDownWaitTimeMinutes": 30,
            "rampDownNotificationMessage": "message",
            "offPeakStartTime": "20:00",
            "offPeakLoadBalancingAlgorithm": "DepthFirst"
```

datr93 commented:
@pilor hey chris can you take another look. We want to add a few more routes and properties to this preview!!

pilor commented:
Please request a new review of the diff from armapireview if you have made changes after approval

datr93 commented:
> Please request a new review of the diff from armapireview if you have made changes after approval

yes I have a email thread with armapireview@microsoft.com and pinged @RamyasreeChakka

datr93 committed:
patch publicNetworkAccess

walrusmcd commented:
thanks for adding the publicnetworkaccess !

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+          "type": "string",
+          "description": "The uri to the storage blob which contains the VHD. Image type must be StorageBlob."
+        },
+        "customId": {
```
cusotmImageId?

datr93 replied:
Its under image object

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+        },
+        "customId": {
+          "type": "string",
+          "description": "The resource id of the custom image or shared image. Image type must be custom."
```
"Image type must be custom." should be "Image type must be CustomImage".

datr93 replied:
ok

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+      }
+    },
+    "DomainInfoProperties": {
+      "description": "Domain configurations of HostPool.",
```
should this be the "Domain configuration of session hosts."?

datr93 replied:
ok

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+                "description": "Agent and other agent side components are delivery schedule is controlled by WVD Infra."
+              },
+              {
+                "value": "TenantAdminControlled",
```
based on the value, Are we forced this property can only be changed by Tenant admin? Should we rename it to "CustomerControlled" or "AdminControllerd"?

datr93 replied:
tenantAdminControlled yes. asked the pm again for you.

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+          "type": "array",
+          "description": "List of secondary maintenance windows. Maintenance windows are 2 hours long. We try to exercise this only when the primary window update fails.",
+          "items": {
+            "$ref": "#/definitions/MaintenanceWindowProperties"
```
The comment we discussed last time is: Primary window will be an object with "hour" property and a single "dayofweek" property.  Secondary windows will be an object with "hour" property and a list of "dayofweek".  Please help to update.

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+      "description": "Maintenance window starting hour and day of week.",
+      "type": "object",
+      "properties": {
+        "hour": {
```
For learning, how do we state a property is required? and for integer type, it is possible to state min and max value?

datr93 replied:
ex:      "required": [
        "hostPoolArmPath",
        "applicationGroupType"
      ],
      
min max I don't think is in a swagger spec for definitions

philipktlin replied:
I found this link, https://swagger.io/docs/specification/data-models/data-types/#range.  Would you please help me if this is supported or share the link of schema definition?  thanks.

datr93 committed:
minor description changes

RamyasreeChakka commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/hasActiveUpdate": {

Why do you need separate API for the validate/progress checks? Can't this information be part of the hostPool GET itself?

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1804 in 3d6e8b1. [](commit_id = 3d6e8b1ce87a31af42513844ba66ef50c33b009c, deletion_comment = False)

RamyasreeChakka commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/initiateUpdate": {

Please take a look at similar scenario in AKS
[https://docs.microsoft.com/en-us/rest/api/aks/agentpools/upgradenodeimageversion](https://docs.microsoft.com/en-us/rest/api/aks/agentpools/upgradenodeimageversion)

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1572 in 3d6e8b1. [](commit_id = 3d6e8b1ce87a31af42513844ba66ef50c33b009c, deletion_comment = False)

RamyasreeChakka commented:
>       "post": {

Consider combining initiateUpdate and validateUpdate to one like below...

POST /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/update
{
 validateOnly: true 
}

Implement as a long-running operation

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1573 in 3d6e8b1. [](commit_id = 3d6e8b1ce87a31af42513844ba66ef50c33b009c, deletion_comment = False)

RamyasreeChakka commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/getUpdate": {

Consider implementing getUpdate & getUpdateProgress as below...

GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses/current
{
}

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1713 in 3d6e8b1. [](commit_id = 3d6e8b1ce87a31af42513844ba66ef50c33b009c, deletion_comment = False)

RamyasreeChakka commented:
@datr93 - As discussed offline, please make the API changes. Let me know after you update this PR with suggested changes.

RamyasreeChakka commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -1569,6 +1569,329 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/validateUpdate/operationResult/{validationInstanceId}": {
```
This URI format is not as per ARM model. Please refer https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#resource-action-requests

datr93 replied:
removed!

datr93 committed:
validate and initiate one endpoint, GET updateStatuses, GET validateStatuses

datr93 commented:
@RamyasreeChakka 

I took the suggestions now:

POST ASYNC: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/update": { validateOnly: "", paramters: "" }

POST ASYNC: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/controlUpdate": { action: "" }

GET:
"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses/current"

GET:
"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/validationStatuses/{validationInstanceId}"

RamyasreeChakka commented:
>       "post": {

As this is a long running operation, define x-ms-long-running-operation and x-ms-long-running-operation-options. Refer [http://azure.github.io/autorest/extensions/#x-ms-long-running-operation](http://azure.github.io/autorest/extensions/#x-ms-long-running-operation)



---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1573 in b9c5d83. [](commit_id = b9c5d83fa08bfcac220cb94b3aef0d1913e38e17, deletion_comment = False)

datr93 committed:
long running and examples

datr93 commented:
> > ```
> >   "post": {
> > ```
> 
> As this is a long running operation, define x-ms-long-running-operation and x-ms-long-running-operation-options. Refer http://azure.github.io/autorest/extensions/#x-ms-long-running-operation
> 
> Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1573 in [b9c5d83](https://github.com/Azure/azure-rest-api-specs/commit/b9c5d83fa08bfcac220cb94b3aef0d1913e38e17). [](commit_id = [b9c5d83](https://github.com/Azure/azure-rest-api-specs/commit/b9c5d83fa08bfcac220cb94b3aef0d1913e38e17), deletion_comment = False)

@RamyasreeChakka 
I added:
        "x-ms-long-running-operation-options": {
          "final-state-via": "location"
        }
For both async calls. Let me know if theres anything else!

shanberg1 commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,6318 @@
+          "$ref": "#/definitions/HostPoolUpdateParameters",
+          "description": "Represents properties for a hostpool update."
+        },
+        "sessionHostConfigurations": {
```
Why plural?  It is one configuration, correct?

datr93 replied:
will change to sessionHostConfiguration

datr93 committed:
rename sessionhostconfigurations -> sessionhostconfigurations, hostpoolupdateparameters -> hostpoolupdateconfiguration

RamyasreeChakka commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses/current": {

As discussed offline, please implement below operations....

Operations results for long running operation
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/operationResults/{operationResultsName} -> follows location 202 pattern
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/operationResults


Proxy resource for long running operation taking >1day
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses/current -> always returns 200
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1672 in 0a97a09. [](commit_id = 0a97a091c1ebe25febd0e54f96390421e96851f7, deletion_comment = False)

datr93 committed:
make updateStatuses/current endpoint a proxy, updateValidationResults, updateResults

datr93 committed:
fix system data and variable

datr93 commented:
@RamyasreeChakka 
I added/changed:

Operations results for long running operation
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateResults/{updateInstanceId} -> follows location 202 pattern
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateValidationResults/{updateValidationInstanceId} -> follows location 202 pattern

Proxy resource for long running operation taking >1day
GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses/current -> always returns 200

shailahirji commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,6389 @@
+            "modelAsString": true,
+            "values": [
+              {
+                "value": "Validating",
```
We dont have a validating state - we had discussed this between you, Sheikh and I @datr93

datr93 replied:
@shailahirji I believe when you start initiate you validate the all of the configurations and hpParams and fail on error right? If not i'll remove it.

datr93 replied:
removed

shailahirji commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,6389 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateStatuses/current": {
```
I dont think this is the right name for this endpoint as it does more than just return the update status for e.g. user might be interested in the shc/ update params and will use the same endpoint @datr93

datr93 replied:
how about updateDetails?

shailahirji commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,6389 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateResults/{updateInstanceId}": {
```
Endpoint duplication? this one and the one above it? @datr93

datr93 replied:
on for validation and one for actual update. we need one for both according to arm team.

datr93 replied:
validate responds with  HostPoolUpdateValidationResponse initiate responds with HostPoolUpdateFullProperties

shailahirji commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -0,0 +1,6389 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/update": {
```
With the proposal of merging initiate and validate - is the flag: ValidateOnly going to be passed in as query string param? @datr93

datr93 replied:
Its passed in with the post:

https://stackoverflow.com/questions/5876809/do-http-post-methods-send-data-as-a-querystring/5876931

datr93 committed:
pr comments rename updateStatuses to updateDetails and remove unsupported state

datr93 committed:
make status endpoints current

RamyasreeChakka commented:
>       "get": {

mark this api as x-ms-pageable

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:30 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

mark this API as x-ms-pageable

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1257 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateValidationResults/current": {

can you rename resource to updateValidationOperationResults?

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1714 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateResults/current": {

can you rename resource to updateOperationResults?

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1763 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

You also need to implement Get collection on /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateDetails

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1673 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

YOu also need to implement GET collection on operation results resource.

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1715 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

YOu also need to implement GET collection on operation results resource.

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1764 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

Why is there no PUT API on this?

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1907 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

mark as x-ms-pageable

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:2689 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

mark x-ms-pageable

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:2875 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>         "tags": [

mark x-ms-pageable

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:2919 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "get": {

mark x-ms-pageable

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:3104 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

RamyasreeChakka commented:
>       "properties": {

Also declare nextLink property here.

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:3172 in b6cdd31. [](commit_id = b6cdd3160217c03b6989c1563f074f7ec95920a4, deletion_comment = False)

datr93 commented:
> 1907

because a session host is created via backend operation. We don't expose it directly

datr93 committed:
pageable and list operation
more next link
more pagination
add refs of private link
more importing issues
undo breaking change
remove unused HostPoolUpdateValidationResponseList
prettier
add description

datr93 committed:
system data for all
examples for all

datr93 commented:
@RamyasreeChakka 
Sessionhost does not have a put because a session host is created via backend operation. We don't expose it directly.
updateValidationOperationResults and updateOperationResults does not have a list operation. I looked at other RPs and they don't have a list, also how would you list operations in 202 async??? There's no body, and listing headers doesn't seem to make sense.

RamyasreeChakka reviewed (approved):
:shipit:

RamyasreeChakka commented:
FYI - Reviewed and approved the changes till below commit/iteration.
1. Commit id: https://github.com/Azure/azure-rest-api-specs/pull/14410/commits/b5e4eea5f8cbf6ba2110c53b613ada3cb5a8d759
2. Iteration: 16 iteration in codeflow.

Not adding 'ARMSignedOff' label as @datr93 team wants to add more changes. @datr93 will follow up with next reviewers for reviewing subsequent commits.

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+      "description": "Session host configurations of HostPool.",
+      "type": "object",
+      "properties": {
+        "vMSizeId": {
```
vmSizeId

datr93 replied:
in all of the swaggers its defined vMSizeId

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+            "values": [
+              {
+                "value": "Gallery",
+                "description": "Using default gallery images offered by azure market place."
```
Azure Marketplace

datr93 replied:
ok

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3601,6 +3609,283 @@
+          "type": "boolean",
+          "description": "Whether to use localTime of the virtual machine."
+        },
+        "maintenanceWindowTimeZone": {
```
I think there is another thread discussing timezone issue, which will impact to scaling plan, agent update and host pool update.

datr93 replied:
ok

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+        "duration": {
+          "type": "integer",
+          "format": "int32",
+          "description": "The duration of the hostpool update in seconds."
```
should it be minutes instead?

datr93 replied:
@Shoeliee

shailahirji replied:
seconds - the admin can opt to send out a maintenance notifications at intervals of seconds.

philipktlin replied:
For maintenance alert duration field, Scott told me that it is only allowed to provide up to minute from portal. And it is not necessary to support in seconds for maintenance alert. would you please help me to understand why customers need to control down to second?

shailahirji replied:
Vidya is OK with minutes as the unit

shailahirji replied:
log off alert - 60 mins is the max that can be configured
maintenance alerts
before hp update kicks off - is between 1 and 5 days
after hp update kicks off - should be > log off alerts and <= 120 mins

philipktlin replied:
Thanks for using minutes as the unit. @datr93 please help to update.

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+        "duration": {
+          "type": "integer",
+          "format": "int32",
+          "description": "The duration of the hostpool update in seconds."
```
The duration of host pool update before/after kick off.

datr93 replied:
@shailahirji is this right?

shailahirji replied:
The time maintenance notification should be sent - relative to before/after the update kickoff

datr93 replied:
The time maintenance notification will be sent relative to before/after the update kickoff in minutes.

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+      "type": "object",
+      "description": "When set schedules the hostpool update at specific time.",
+      "properties": {
+        "time": {
```
in what format? UTC ? ISO-8601?

philipktlin replied:
should it be "datetime"?

datr93 replied:
> in what format? UTC ? ISO-8601?

https://swagger.io/docs/specification/data-models/data-types/#string
date-time – the date-time notation as defined by RFC 3339, section 5.6, for example, 2017-07-21T17:32:28Z

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+            "InProgress",
+            "Pausing",
+            "Paused",
+            "StartFailed",
```
When/what situation to use "StartFailed"?

datr93 replied:
@shailahirji

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
-        "exactVersion": {
-          "type": "string",
-          "description": "The exact version of the image."
+        "listOfErrors": {
```
just use "errors"

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+          "format": "float",
+          "description": "The progress percentage."
+        },
+        "sessionHostsToMigrate": {
```
why use "Migrate" wording? should we use "sessionHostsToUpdate", "sessionHostsToUpdating", and "sessionHostsToUpdated"?

datr93 replied:
@shailahirji they came up with the names

shailahirji replied:
ok to substitute Migrate with HostPoolUpdate

philipktlin replied:
do you mean "update", e.g. "sessionHostsToMigrate" => "sessionHostsToUpdate"?

philipktlin replied:
@datr93 please help to find all "migrate" usage (in any form, e.g. migration or migrated, etc.) to replace with "update".

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+          "format": "int32",
+          "description": "The number of session hosts migrated by hostpool update."
+        },
+        "sessionHostsRollbackFailed": {
```
Just use "sessionHostsFailed"; "sessionHostsRollbackFailed" is confusing, does it mean rollback failed or update failed?

datr93 replied:
@shailahirji they came up with the names

shailahirji replied:
rollback failed

philipktlin replied:
Should we rename it to "sessionHostsFailed"?

datr93 replied:
so keep? because it gives more infor this is for when a rollback has failed

philipktlin replied:
To clarify, we have numbers for "sessionHostsToUpdate", "sessionHostsUpdating", "sessionHostsUpdated", and "sessionHostsRollbackFailed".  But sum of "sessionHostsUpdating", "sessionHostsUpdated", and "sessionHostsRollbackFailed" is not equal to "sessionHostsToUpdate" because we don't have number for "sessionHostsFailed"/"sessionHostsUpdateFailed". add this number? @shailahirji what do you think?

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -1569,6 +1569,329 @@
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/hasActiveUpdate": {
```
why do we need this method? if it can be fulfilled by HostpoolUpdate Get or HostpoolUpdateProgress Get methods.

datr93 replied:
outdated we don't have this anymore update team wanted it originally.

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+          "type": "string",
+          "description": "The uri to the storage blob which contains the VHD. Image type must be StorageBlob."
+        },
+        "customId": {
```
customImageId

datr93 replied:
its already nested under ImageInfo so I omitted image form the name

datr93 committed:
move shconfig to proxy obj

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+        }
+      }
+    },
+    "KeyvaultCredentialProperties": {
```
key vault is 2 words, therefore it should be "KeyVaultCredentialProperties"? btw I found some properties start with capital and some are not, would you please help me to understand?

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+            "$ref": "#/definitions/MaintenanceAlertsProperties"
+          }
+        },
+        "logOffDelaySeconds": {
```
I would like to ask log off message is PC message, which is a message popped up on VM screen, right? if so, should we have a maintenaceMessages property as array to define both logoff message and cancel update message.  Also I think we should group logOffDelaySeconds and logOffMessage together.

datr93 replied:
@shailahirji

shailahirji replied:
not following what cancel update message means here "logoff message and cancel update message". By grouping if you mean create an object that contains logOffDelaySeconds  and logOffMessage as properties -sure

philipktlin replied:
Sheikh suggested during last meeting that when customer cancel host pool update, we would like to support to send cancel host pool update meeting like logoff message.

> By grouping if you mean create an object that contains logOffDelaySeconds and logOffMessage as properties -sure
yes.

philipktlin replied:
My suggestion: 
{
   "sessionHostMessages": [
      { "type": "logOff", "timeout": "30", "message": "xxxxxxx" }
      { "type": "cancelUpdate", "timeout": "30", "message": "xxxxxxx" }
   ]
}

Reference: I found the wording "Timeout" from https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/shutdown

datr93 committed:
add cancel message

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+        },
+        "sessionHostConfigurations": {
+          "$ref": "#/definitions/SessionHostConfigurationProperties",
+          "description": "The session host configurations of HostPool."
```
"The session host configurations of a host pool."

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -3637,181 +3966,496 @@
+        },
+        "hostPoolUpdateParameters": {
+          "$ref": "#/definitions/HostPoolUpdateParameters",
+          "description": "Represents properties for a hostpool update."
```
Remove "Represents" wording

philipktlin commented on specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:
```diff
@@ -4858,6 +5527,76 @@
+        "sessionHostConfigurationLastUpdateTime": {
+          "type": "string",
+          "format": "date-time",
+          "description": "This time will match the time in the SHC for when the update was initiated.",
```
should we use "SHC" alias or "session host configuration"?

datr93 committed:
fix formatting

datr93 committed:
add parameter

datr93 committed:
philips pr comments

datr93 committed:
fix examples

datr93 committed:
move lastUpdate to shc proxy rename to version

datr93 committed:
pr comments migrate to update

datr93 committed:
add object type to private link

datr93 committed:
remove object type for one

chiragg4u commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DesktopVirtualization/hostPools/{hostPoolName}/updateDetails/singleton": {

We don't have singleton pattern; we've many places where RP's have used current or default. Any reason for defining it using singleton? Can't we use default?

---
Refers to: specification/desktopvirtualization/resource-manager/Microsoft.DesktopVirtualization/preview/2021-05-13-preview/desktopvirtualization.json:1957 in 619aea2. [](commit_id = 619aea25ff468ce4627081ed28e910bad94c4e5e, deletion_comment = False)

datr93 committed:
change to default

chiragg4u commented:
Reviewed the changes from last approval, it looks good. Please follow the process given @ https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/212/Swagger-PR-Review

datr93 commented:
@zhenglaizhang can you take a look!

zhenglaizhang commented:
Hi @datr93 pls help fix the LintDiff warning in https://github.com/Azure/azure-rest-api-specs/pull/14410#issuecomment-840823254, it's generally adviced to fix them fow new API version.

datr93 commented:
> Hi @datr93 pls help fix the LintDiff warning in [#14410 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/14410#issuecomment-840823254), it's generally adviced to fix them fow new API version.


the UpdateValidationOperationResults_Get is not returning an array
its an object with two arrays inside
and the booleans make sense to be booleans

zhenglaizhang commented:
> > Hi @datr93 pls help fix the LintDiff warning in [#14410 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/14410#issuecomment-840823254), it's generally adviced to fix them fow new API version.
> 
> Warmings isn't applicable
> the UpdateValidationOperationResults_Get is not returning an array
> its an object with two arrays inside
> and the booleans make sense to be booleans

It makes sense, all LGTM, so if you are fine, I will proceed to merge the PR

datr93 commented:
@zhenglaizhang please merge!!!!
